common/libs/contract/ABIFunction.js

```diff
@@ -0,0 +1,160 @@
+import abi from 'ethereumjs-abi';
+import { toChecksumAddress } from 'ethereumjs-util';
+import BigNumber from 'bignumber.js';
+
+export default class AbiFunction {
+  constant;
+  funcParams;
+  inputs;
+  inputNames;
+  inputTypes;
+  outputs;
+  outputNames;
+  outputTypes;
+  methodSelector;
+  name;
+  payable;
+  type;
+
+  constructor(abiFunc, outputMappings) {
+    Object.assign(this, abiFunc);
+    this._init(outputMappings);
+  }
+
+  _init(outputMappings = []) {
+    const { inputs, outputs } = this;
+    this.funcParams = this._makeFuncParams();
+    //TODO: do this in O(n)
+    this.inputTypes = inputs.map(({ type }) => type);
+    this.outputTypes = outputs.map(({ type }) => type);
+    this.inputNames = inputs.map(({ name }) => name);
+    this.outputNames = outputs.map(
+      ({ name }, i) => outputMappings[i] || name || i
+    );
+
+    this.methodSelector = abi
+      .methodID(this.name, this.inputTypes)
+      .toString('hex');
+  }
+
+  encodeInput = (suppliedInputs = {}) => {
```

Comment:
I know flow typing is TBD which should clear this up, but I don't think we should use default args here, I'd want this to fail loud if the wrong thing is sent.

Reply:
Reason why default arg is an object here is because `_processSuppliedArgs` does type/error checking per object key, which allows for more specific errors on what the supplied input is missing or what specific key is malformed.